modules: Add `_prefix` module argument, improve error, add docs (#398839)
NixOS/nixpkgs: doc/module-system/module-system.chapter.md

#### `class` {#module-system-lib-evalModules-return-value-_configurationClass}

The [`class` argument](#module-system-lib-evalModules-param-class).

## Module arguments {#module-system-module-arguments}

Module arguments are the attribute values passed to modules when they are evaluated.

They originate from these sources:
1. Built-in arguments
  - `lib`,
  - `config`,
  - `options`,
  - `_class`,
  - `_prefix`,

> 2. Attributes from the [`specialArgs`] which were passed to [`evalModules`] or `submoduleWith`. These are application-specific.

Reviewer: ```suggestion
2. Attributes from the [`specialArgs`] argument passed to [`evalModules`] or `submoduleWith`. These are application-specific.
```

3. Attributes from the `_module.args` option value. These are application-specific and can be provided by any module.

The prior two categories are available while evaluating the `imports`, whereas
the last category is only available after the `imports` have been resolved.

> ### `lib` {#module-system-module-argument-lib}

Author: (anchors useless)

Just anchors isn't good enough because there isn't any visual indication that they exist.
I've worked around it.

This should be automated in the tooling.

A reference to the Nixpkgs library.

### `config` {#module-system-module-argument-config}

All option values.
Unlike the `evalModules` [`config` return attribute](#module-system-lib-evalModules-return-value-config), this includes `_module`.

### `options` {#module-system-module-argument-options}

All evaluated option declarations.

### `_class` {#module-system-module-argument-_class}

The [expected class](#module-system-lib-evalModules-param-class) of the loaded modules.

### `_prefix` {#module-system-module-argument-_prefix}

> The location under which the module is evaluated. This is used to improve error reporting and find the implicit `name` module argument in submodules.

Reviewer: ```suggestion
The location under which the module is evaluated.
This is used to improve error reporting and to find the implicit `name` module argument in submodules.
It is exposed as a module argument due to how the module system is implemented, which cannot be avoided without breaking compatibility.
```

I think it's more important to know what's in the PR description: that it's merely an artifact of module system internals, that happens to spill out.

It is a good practice not to rely on `_prefix`. A module should not make assumptions about its location in the configuration tree.
For example, the root of a NixOS configuration may have a non-empty prefix, for example when it is a specialisation, or when it is part of a larger, multi-host configuration such as a [NixOS test](https://nixos.org/manual/nixos/unstable/#sec-nixos-tests).

> Any dependencies on `_prefix` should be replaced with explicit options, whose default definitions can be provided by the module that imports them.

Reviewer: ```suggestion
Instead of depending on `_prefix` use explicit options, whose default definitions can be provided by the module that imports them.
```

You're writing it as if there already were dependencies :)
